### PR TITLE
fix #18079 Illegal storage access compiling call with nested ref/dere…

### LINKED ISSUE #18079: Illegal storage access compiling call with nested ref/deref'd types
Calling a function with nested layers of object construction, each of which constructs a reference object then dereferences it, results in `SIGSEGV` at compile time with Nim versions 1.2.12, 1.4.2, and current git head.

### Example
```nim
type
  Foo = object
    discard

  Bar = object
    x: Foo

proc baz(state: var Bar) = discard
baz((ref Bar)(x: (new Foo)[])[])
```

### Current Output
```
$ nim c sigsegv.nim
Hint: used config file '/etc/nim/nim.cfg' [Conf]
Hint: used config file '/etc/nim/config.nims' [Conf]
....SIGSEGV: Illegal storage access. (Attempt to read from nil?)
$ nim --version
Nim Compiler Version 1.4.2 [Linux: amd64]
Compiled at 2021-05-13
Copyright (c) 2006-2020 by Andreas Rumpf

active boot switches: -d:release
```
```
$ ~/Nim/bin/nim c sigsegv.nim
Hint: used config file 'Nim/config/nim.cfg' [Conf]
Hint: used config file 'Nim/config/config.nims' [Conf]
....SIGSEGV: Illegal storage access. (Attempt to read from nil?)
Segmentation fault
$ ~/Nim/bin/nim --version
Nim Compiler Version 1.5.1 [Linux: amd64]
Compiled at 2021-05-24
Copyright (c) 2006-2021 by Andreas Rumpf

git hash: d217888e5679aff063668930bd00892c4f5cb2b3
active boot switches: -d:release
```
```
$ ./env.sh nim c ../sigsegv.nim 
Hint: system [Processing]
Hint: widestrs [Processing]
Hint: io [Processing]
Hint: sigsegv [Processing]
SIGSEGV: Illegal storage access. (Attempt to read from nil?)
$ ./env.sh nim --version
Nim Compiler Version 1.2.12 [Linux: amd64]
Compiled at 2021-05-23
Copyright (c) 2006-2020 by Andreas Rumpf

git hash: 121628357ec7fae91335bd392f03b0e6dd249319
active boot switches: -d:release
```

### Expected Output
Not a segfault.

### PR DESCRIPTION
**Pull request by bung87**
…f types

fix #18079

### CODE REVIEW

File: compiler/ccgexprs.nim [hunk 1]
```diff
@@ -1398,7 +1398,7 @@ proc rawGenNew(p: BProc, a: var TLoc, sizeExpr: Rope; needsInit: bool) =
   var b: TLoc
   initLoc(b, locExpr, a.lode, OnHeap)
   let refType = typ.skipTypes(abstractInstOwned)
-  assert refType.kind == tyRef
+  # assert refType.kind == tyRef
```
Comment (Araq) [inline compiler/ccgexprs.nim:1401]: No.

File: compiler/ccgexprs.nim [hunk 2]
```diff
@@ -1551,11 +1552,11 @@ proc genObjConstr(p: BProc, e: PNode, d: var TLoc) =
     ]#
     if handleConstExpr(p, e, d): return
   var t = e.typ.skipTypes(abstractInstOwned)
-  let isRef = t.kind == tyRef
+  let isRefLike = t.kind == tyRef or t.kind == tyVar
```
Comment (Araq) [inline compiler/ccgexprs.nim:1555]: Nah.

File: compiler/ccgexprs.nim [hunk 3]
```diff
@@ -1589,10 +1591,10 @@ proc genObjConstr(p: BProc, e: PNode, d: var TLoc) =
     tmp2.r.add(field.loc.r)
     if useTemp:
       tmp2.k = locTemp
-      tmp2.storage = if isRef: OnHeap else: OnStack
+      tmp2.storage = if isRefLike: OnHeap else: OnStack
```
Comment (Araq) [inline compiler/ccgexprs.nim:1594]: For tyVar you don't know if it's on the heap or not.

**Comment by Araq** [comment]
The real fix is probably that one cannot optimize away nkHiddenAddr(nkDeref(x)) as the typing rules don't work out. It's optimized in transf.nim iirc.

**Comment by bung87** [comment]
> The real fix is probably that one cannot optimize away nkHiddenAddr(nkDeref(x)) as the typing rules don't work out. It's optimized in transf.nim iirc.

okay, I'll try.

**Pull request closed without merging**